crates/red_knot_workspace/tests/check.rs

```diff
@@ -1,48 +1,114 @@
-        pull_types(&db, file);
-
-        // try the file as a stub also
-        println!("re-checking as .pyi");
-        let pyi_dest = root.join(format!("{source_fn}i"));
-        std::fs::copy(source, pyi_dest.as_std_path())?;
-        let file = system_path_to_file(&db, pyi_dest).unwrap();
-        pull_types(&db, file);
+    let root = SystemPathBuf::from("/src");
+
+    let system = TestSystem::default();
+    let memory_fs = system.memory_file_system();
+    memory_fs.create_directory_all(root.as_ref())?;
+
+    let mut db = setup_db(&root, system.clone())?;
+
+    let crate_root = String::from(env!("CARGO_MANIFEST_DIR"));
+    let workspace_root = get_workspace_root()?;
+    let workspace_root = workspace_root.to_string();
+
+    let corpus = vec![
+        format!("{crate_root}/resources/test/corpus/**/*.py"),
+        format!("{workspace_root}/crates/ruff_python_parser/resources/**/*.py"),
+        format!("{workspace_root}/crates/ruff_linter/resources/**/*.py"),
+        // TODO: Enable running over typeshed stubs once there are fewer failures:
+        // format!("{workspace_root}/crates/red_knot_vendored/vendor/typeshed/**/*.pyi"),
+    ]
+    .into_iter()
+    .flat_map(|pattern| glob::glob(&pattern).unwrap());
+
+    for path in corpus {
+        let path = path?;
+        let relative_path = path.strip_prefix(&workspace_root)?;
+
+        let (py_expected_to_fail, pyi_expected_to_fail) = KNOWN_FAILURES
+            .iter()
+            .find_map(|(path, py_fail, pyi_fail)| {
+                if Some(*path) == relative_path.to_str() {
+                    Some((*py_fail, *pyi_fail))
+                } else {
+                    None
+                }
+            })
+            .unwrap_or((false, false));
+
+        let source = path.as_path();
+        let source_filename = source.file_name().unwrap().to_str().unwrap();
+
+        let code = std::fs::read_to_string(source)?;
+
+        let mut check_with_file_name = |path: &SystemPath| {
+            memory_fs.remove_all();
+            memory_fs.write_file(path, &code).unwrap();
+            File::sync_path(&mut db, path);
+
+            // this test is only asserting that we can pull every expression type without a panic
+            // (and some non-expressions that clearly define a single type)
+            let file = system_path_to_file(&db, path).unwrap();
+
+            let result = std::panic::catch_unwind(|| pull_types(&db, file));
+
+            let expected_to_fail = if path.extension().map(|e| e == "pyi").unwrap_or(false) {
+                pyi_expected_to_fail
+            } else {
+                py_expected_to_fail
+            };
+            if let Err(err) = result {
+                if !expected_to_fail {
+                    std::panic::resume_unwind(err);
+                }
+            } else {
+                assert!(!expected_to_fail, "Expected to panic, but did not. Consider removing this path from KNOWN_FAILURES");
+            }
```

Comment:
We now risk to have a stale file around when `file` gets deleted. 
```suggestion
            memory_fs.write_file(path, &code).unwrap();
            File::sync_path(&mut db, path);

            // this test is only asserting that we can pull every expression type without a panic
            // (and some non-expressions that clearly define a single type)
            let file = system_path_to_file(&db, path).unwrap();

            let result = std::panic::catch_unwind(|| pull_types(&db, file));

            let expected_to_fail = if path.extension().map(|e| e == "pyi").unwrap_or(false) {
                pyi_expected_to_fail
            } else {
                py_expected_to_fail
            };
            if let Err(err) = result {
                if !expected_to_fail {
                    std::panic::resume_unwind(err);
                }
            } else {
                assert!(!expected_to_fail, "Expected to panic, but did not. Consider removing this path from KNOWN_FAILURES");
            }
            
            memory_fs.remove_all();
            file.sync(db);
```